fishtank/src/backend/docker.ts

```diff
@@ -0,0 +1,88 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
+import { ChildProcess, execFile, ExecFileOptions } from 'child_process'
+import { promisify } from 'util'
+
+const DEFAULT_COMMAND = 'docker'
+
+type ExecFilePromiseReturn = { child: ChildProcess; stdout: string; stderr: string }
+
+type ExecFilePromiseError = {
+  cmd: string
+  code: number | string
+  stdout: string
+  stderr: string
+}
+
+const execChildProcess = promisify<
+  string,
+  readonly string[],
+  ExecFileOptions,
+  ExecFilePromiseReturn
+>(execFile)
```

Comment:
How do you decide on execFile vs exec. I read the [difference](https://nodejs.org/api/child_process.html#child_processexecfilefile-args-options-callback) 

> The child_process.execFile() function is similar to [child_process.exec()](https://nodejs.org/api/child_process.html#child_processexeccommand-options-callback) except that it does not spawn a shell by default. Rather, the specified executable file is spawned directly as a new process making it slightly more efficient than [child_process.exec()](https://nodejs.org/api/child_process.html#child_processexeccommand-options-callback).
> 
> The same options as [child_process.exec()](https://nodejs.org/api/child_process.html#child_processexeccommand-options-callback) are supported. Since a shell is not spawned, behaviors such as I/O redirection and file globbing are not supported.

Is it because execFile is fast? Is it okay with the tradeoff? execFile is not good if we want to consume the data in real time from the child process,  plus it does not support Windows because no shell.

Reply:
It has been chosen for performance, and for security (to avoid shell injection).

It can be used to grab process output. The documentation you linked says that explicitly:

> The stdout and stderr arguments passed to the callback will contain the stdout and stderr output of the child process

This can be easily verified with this minimal, reproducible example:

```typescript
import { ChildProcess, execFile, ExecFileOptions } from 'child_process'
import { promisify } from 'util'

type ExecFilePromiseReturn = { child: ChildProcess; stdout: string; stderr: string }

const execChildProcess = promisify<
  string,
  readonly string[],
  ExecFileOptions,
  ExecFilePromiseReturn
>(execFile)

execChildProcess('echo', ['-n', 'hello'], {}).then(({ stdout }) => {
  console.log(`The output from execFile was: "${stdout}"`)
})
```

What the excerpt of documentation you quoted says is that I/O redirection is not supported, which is a different feature that we don't need (and, if we ever need it, it would be more safely implemented in TypeScript rather than using a shell).

The documentation also explicitly says that `execFile` *is* supported on Windows, only `.bat` and `.cmd` files are not *directly* supported. But actually even `execFile` can spawn `.bat` and `.cmd` files simply by invoking `cmd.exe` directly.

Lastly, note that `exec()` is built upon `execFile()`, so anything that `exec()` can do, `execFile()` can do it as well.

Reply:
> What the excerpt of documentation you quoted says is that I/O redirection is not supported
The Nodejs doc I linked above, it says that.

When you say we don't need that, what is your thought process and reasoning? And do you mean we dont need that for the basic skeleton stage or hardfork testing?

Reply:
I/O redirection is that feature that allows you to pipe output to files or other program, e.g.:

```
echo hello > file
echo hello | cat
```

That's the syntax that would not be supported by `execFile`, because that's a shell-specific syntax (more specifically, it's a Unix shell feature; not sure if it would work on Windows). The same functionality can be implemented using `execFile` too, the only difference is that you need to write TypeScript code instead of shell code, so no `|` and `>`, but other constructs.

I do not think we'll need this functionality (whether it's written in shell scripting or in TypeScript) because I do not see a reason to pipe output into other programs or files. I do not see how any of the goals of this project would require using any of these I/O redirection features.

Reply:
Thanks for sharing your reasoning. The requirements in the spec do required this, but we talked about it during spec meeting. One testing case we want to do is to verify blocks are rejected after activation if timestamp is invalid. The rejection message is in the mining logs. Do you plan to parse the stdout to achieve it? the mining logs can be long and might be hard to process without saving to file. Not saying we should not use `execFile`, but just want to point it out to make sure we have thorough and informative consideration when make the decision.

Reply:
A separate thing somewhat related to this, do we need the streaming functionality for logs? like existing simulator does with `ChildProcessWithoutNullStreams`. What is your thoughts and plan on the streams of `ironfish start` and `ironfish miner:start` cmd?

Reply:
I don't know exactly the specifics of how we will parse those logs, but by looking around I see we can use [`node/getLogStream` from the RPC](https://ironfish.network/developers/documentation/rpc/node/get_log_stream) to get node logs. For miner logs, we may need to stream stdout/stderr as you are suggesting. We will come up with a solution for that problem once we are at that stage.

The fact that the current code makes use of a specific variant of `execFile` does not mean that all the code we write in the future will have to use that. Different problems may have different solution, and as such we may end up using more than 1 variant of `execFile`, and maybe some other functions provided by `child_process`, like `spawn`. For the code in this PR, `execFile` is sufficient and does the job that this PR is trying to accomplish. Future code will make use of whatever function it needs.

Reply:
The plan you lay out in above sounds good to me. I like the idea of keeping the exec as an option and might use it for streams. Last ask, can you add the miner logs in the spec requirement and create a linear task for it?

Reply:
Created linear task https://linear.app/if-labs/issue/IFL-1628/ability-to-storeconsume-the-node-logs 
Updated the spec requirement.